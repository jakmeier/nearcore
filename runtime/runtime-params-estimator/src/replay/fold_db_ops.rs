use super::cache_stats::CacheStats;
use super::Visitor;
use std::collections::BTreeMap;
use std::io::Write;

const EMPTY_STATE: &str = "states must never be empty";

/// A visitor that keeps track of DB operations and aggregates it by specific labels.
pub(super) struct FoldDbOps {
    /// Labels at which to break aggregation.
    fold_anchors: Vec<String>,
    /// Fields to display on each output, if they are available in the trace.
    printed_fields: Vec<String>,
    /// Print the DB operations that are at indentation 0, not associated with a span.
    print_top_level: bool,
    /// Only show data for a specific smart contract, specified by account id.
    /// Currently only applies within receipts.
    account_filter: Option<String>,
    /// Skip all input lines that have deeper indentation.
    skip_indent: Option<usize>,
    /// Optionally collect and print detailed statistics for cache hits and misses.
    track_caches: bool,
    /// Keeps track of current block.
    block_hash: Option<String>,
    /// Stack of states.
    states: Vec<State>,
}

#[derive(Default)]
struct State {
    /// The indent at which this state started
    indent: usize,
    /// Keeps track of operations per DB column.
    ops_cols: BTreeMap<String, BTreeMap<String, usize>>,
    /// Optionally collect and print detailed statistics for cache hits and misses.
    cache_stats: Option<CacheStats>,
}

impl FoldDbOps {
    pub(super) fn new() -> Self {
        Self {
            fold_anchors: vec![],
            printed_fields: vec![],
            print_top_level: true,
            account_filter: None,
            track_caches: false,
            states: vec![State::default()],
            block_hash: None,
            skip_indent: None,
        }
    }

    /// Pre-set that folds on chunks.
    pub(super) fn chunks(self) -> Self {
        self.fold("apply_transactions").print_field("receiver").print_field("shard_id")
    }

    /// Pre-set that folds on receipts.
    pub(super) fn receipts(self) -> Self {
        self.fold("process_receipt")
            .print_field("receiver")
            .print_field("receipt_id")
            .fold("process_transaction")
            .print_field("receipt_id")
            .print_top_level(false)
    }

    pub(super) fn fold(mut self, anchor: impl Into<String>) -> Self {
        self.fold_anchors.push(anchor.into());
        self
    }

    pub(super) fn print_field(mut self, field: impl Into<String>) -> Self {
        self.printed_fields.push(field.into());
        self
    }

    pub(super) fn print_top_level(mut self, yes: bool) -> Self {
        self.print_top_level = yes;
        self
    }

    pub(super) fn with_cache_stats(mut self) -> Self {
        self.track_caches = true;
        self.state().cache_stats = Some(CacheStats::default());
        self
    }

    pub(super) fn account_filter(mut self, account: Option<String>) -> Self {
        self.account_filter = account;
        self
    }

    fn state(&mut self) -> &mut State {
        self.states.last_mut().expect(EMPTY_STATE)
    }

    fn push_state(&mut self, indent: usize) {
        let cache_stats = if self.track_caches { Some(CacheStats::default()) } else { None };
        let new_state = State { indent, ops_cols: Default::default(), cache_stats };
        self.states.push(new_state);
    }

    fn pop_state(&mut self) -> State {
        let state = self.states.pop().expect(EMPTY_STATE);
        if self.states.is_empty() {
            self.push_state(0);
        }
        state
    }

    fn skip(&mut self, indent: usize) -> bool {
        if let Some(skip_indent) = self.skip_indent {
            if skip_indent >= indent {
                self.skip_indent = None;
            } else {
                return true;
            }
        }
        false
    }

    /// Check if indentation has gone back enough to pop current state.
    fn check_indent(&mut self, out: &mut dyn Write, indent: usize) -> anyhow::Result<()> {
        if self.states.len() > 1 && self.state().indent >= indent {
            self.pop_state().print(out)?;
        }
        Ok(())
    }
}

impl Visitor for FoldDbOps {
    fn eval_db_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        size: Option<u64>,
        key: &[u8],
        col: &str,
    ) -> anyhow::Result<()> {
        // Block hash in traces is visibly by looking at DB lookups of
        // `BlockInfo` column. Keep track of it so that each time something is
        // printed, the block hash can be included if desired.
        match op {
            "GET" => {
                if col == "BlockInfo" {
                    self.block_hash = Some(bs58::encode(key).into_string());
                }
            }
            _ => {
                // nop
            }
        }

        if self.skip(indent) {
            return Ok(());
        }

        // Count DB operation for the corresponding column.
        *self
            .state()
            .ops_cols
            .entry(op.to_owned())
            .or_default()
            .entry(col.to_owned())
            .or_default() += 1;

        if let Some(cache_stats) = &mut self.state().cache_stats {
            cache_stats.eval_db_op(op, size);
        }

        self.check_indent(out, indent)?;
        Ok(())
    }

    fn eval_storage_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        if self.skip(indent) {
            return Ok(());
        }
        if let Some(cache_stats) = &mut self.state().cache_stats {
            cache_stats.eval_storage_op(op, dict)?;
        }
        self.check_indent(out, indent)?;
        Ok(())
    }

    fn eval_label(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        label: &str,
        dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        if self.skip(indent) {
            return Ok(());
        }
        self.check_indent(out, indent)?;
        if let (Some(receiver), Some(filtered_account)) =
            (dict.get("receiver"), &self.account_filter)
        {
            if receiver != filtered_account {
                self.skip_indent = Some(indent);
                return Ok(());
            }
        }
        if self.fold_anchors.iter().any(|anchor| *anchor == label) {
            // Section to fold on starts. Push a new state on the stack and
            // print the header for the new section.
            self.push_state(indent);
            write!(out, "{:indent$}{label}", "")?;
            for key in self.printed_fields.iter() {
                if let Some(value) = dict.get(key.as_str()) {
                    write!(out, " {key}={value}")?;
                }
            }
            if let Some(block) = &self.block_hash {
                write!(out, " block={block}")?;
            }
            writeln!(out)?;
        }

        if let Some(cache_stats) = &mut self.state().cache_stats {
            cache_stats.eval_generic_label(dict);
        }
        Ok(())
    }

    fn flush(&mut self, out: &mut dyn Write) -> anyhow::Result<()> {
        if self.print_top_level {
            writeln!(out, "top-level:")?;
            self.pop_state().print(out)?;
        }
        Ok(())
    }
}

impl State {
    fn print(self, out: &mut dyn Write) -> anyhow::Result<()> {
        let indent = self.indent + 2;
        for (op, map) in self.ops_cols.into_iter() {
            if !map.is_empty() {
                write!(out, "{:indent$}{op}   ", "")?;
            }
            for (col, num) in map.into_iter() {
                write!(out, "{num:8>} {col}  ")?;
            }
            writeln!(out)?;
        }

        if let Some(stats) = self.cache_stats {
            stats.print(out, indent)?;
        }
        writeln!(out)?;
        Ok(())
    }
}
