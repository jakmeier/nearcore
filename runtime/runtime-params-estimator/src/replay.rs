use anyhow::Context;
use std::collections::BTreeMap;
use std::fs::File;
use std::io::{self, BufRead, Write};
use std::path::PathBuf;
use std::str::SplitWhitespace;
use tracing::log::error;

use self::fold_db_ops::FoldDbOps;
use self::gas_charges::ChargedVsFree;

mod cache_stats;
mod fold_db_ops;
mod gas_charges;

#[derive(clap::Parser)]
pub(crate) struct ReplayCmd {
    trace: PathBuf,
    #[clap(subcommand)]
    mode: ReplayMode,
    /// Only show data for a specific smart contract, specified by account id.
    #[clap(long)]
    account: Option<String>,
}

#[derive(Clone, Copy, clap::Subcommand, Debug)]
pub(crate) enum ReplayMode {
    /// Print DB accesses and cache statistics for the entire trace.
    CacheStats,
    /// Print DB accesses per receipt.
    ReceiptDbStats,
    /// Print DB accesses and cache statistics per receipt.
    ReceiptCacheStats,
    /// Print DB accesses per chunk.
    ChunkDbStats,
    /// Print DB accesses and cache statistics per chunk.
    ChunkCacheStats,
    /// Go over DB operations and print how much of it is paid for with gas.
    GasCharges,
}

impl ReplayCmd {
    pub(crate) fn run(&self, out: &mut dyn Write) -> anyhow::Result<()> {
        let file = File::open(&self.trace)?;

        let mut visitor = self.build_visitor();

        for line in io::BufReader::new(file).lines() {
            let line = line?;
            if let Err(e) = visitor.eval_line(out, &line) {
                error!("ERROR: {e} for input line: {line}");
            }
        }
        visitor.flush(out)?;

        Ok(())
    }

    fn build_visitor(&self) -> Box<dyn Visitor> {
        match &self.mode {
            ReplayMode::CacheStats => {
                Box::new(FoldDbOps::new().with_cache_stats().account_filter(self.account.clone()))
            }
            ReplayMode::ChunkDbStats => {
                if self.account.is_some() {
                    unimplemented!("account filter does not work with per-chunk statistics");
                }
                Box::new(FoldDbOps::new().chunks())
            }
            ReplayMode::ChunkCacheStats => {
                if self.account.is_some() {
                    unimplemented!("account filter does not work with per-chunk statistics");
                }
                Box::new(FoldDbOps::new().chunks().with_cache_stats())
            }
            ReplayMode::ReceiptDbStats => {
                Box::new(FoldDbOps::new().receipts().account_filter(self.account.clone()))
            }
            ReplayMode::ReceiptCacheStats => Box::new(
                FoldDbOps::new().receipts().with_cache_stats().account_filter(self.account.clone()),
            ),
            ReplayMode::GasCharges => {
                if self.account.is_some() {
                    unimplemented!("account filter does not work with gas charges");
                }
                Box::new(ChargedVsFree::default())
            }
        }
    }
}

/// Interface for processing an IO trace line-by-line.
///
/// The visitor default methods take over the basic parsing of the IO trace.
/// Specific implementations can extract the data of interest and process it
/// further as necessary, without duplicating the parsing code.
trait Visitor {
    /// The root entry point of the visitors.
    ///
    /// This function takes a raw input line as input without any preprocessing.
    /// A visitor may choose to overwrite this function for full control but the
    /// intention is that the default implementation takes over the basic
    /// parsing and visitor implementations define their behaviour using the
    /// other trait methods.
    fn eval_line(&mut self, out: &mut dyn Write, line: &str) -> anyhow::Result<()> {
        if let Some(indent) = line.chars().position(|c| !c.is_whitespace()) {
            let mut tokens = line.split_whitespace();
            if let Some(keyword) = tokens.next() {
                match keyword {
                    "GET" | "SET" | "UPDATE_RC" => {
                        let col = tokens.next().context("missing column field in DB operation")?;
                        let mut key_str = tokens.next().context("missing key in DB operation")?;
                        if key_str.starts_with('"') {
                            key_str = &key_str[1..key_str.len() - 1];
                        }
                        let key = bs58::decode(key_str).into_vec()?;
                        let dict = extract_key_values(tokens)?;
                        let size: Option<u64> = dict.get("size").map(|s| s.parse()).transpose()?;
                        self.eval_db_op(out, indent, keyword, size, &key, col)?;
                    }
                    "storage_read" | "storage_write" | "storage_remove" | "storage_has_key" => {
                        let op = tokens.next();
                        if op.is_none() {
                            return Ok(());
                        }

                        let dict = extract_key_values(tokens)?;
                        self.eval_storage_op(out, indent, keyword, &dict)?;
                    }
                    other_label => {
                        let dict = extract_key_values(tokens)?;
                        self.eval_label(out, indent, other_label, &dict)?;
                    }
                }
            }
        }
        Ok(())
    }

    /// Gets called for every trie storage operation.
    ///
    /// A storage operation is a layer above DB operations, one storage
    /// operation can cause many DB operations, or it could be resolved from
    /// cache without any DB operations.
    fn eval_storage_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        let (_, _, _, _) = (out, indent, op, dict);
        Ok(())
    }

    /// Gets called for every DB operation.
    fn eval_db_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        size: Option<u64>,
        key: &[u8],
        col: &str,
    ) -> anyhow::Result<()> {
        if col == "State" {
            self.eval_state_db_op(out, indent, op, size, key)
        } else {
            Ok(())
        }
    }

    /// Gets called for every DB operation on the state column.
    fn eval_state_db_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        size: Option<u64>,
        key: &[u8],
    ) -> anyhow::Result<()> {
        let (_, _, _, _, _) = (out, indent, op, size, key);
        Ok(())
    }

    /// Opening spans that are not storage or DB operations.
    fn eval_label(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        label: &str,
        dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        let (_, _, _, _) = (out, indent, label, dict);
        Ok(())
    }

    /// Print accumulated internal state.
    ///
    /// This function is called at the end of a replayed trace. Visitor
    /// implementation usually choose to call this more often. For example, once
    /// for each receipt.
    fn flush(&mut self, out: &mut dyn Write) -> anyhow::Result<()> {
        let _ = out;
        Ok(())
    }
}

fn extract_key_values<'a>(
    mut tokens: SplitWhitespace<'a>,
) -> anyhow::Result<BTreeMap<&'a str, &'a str>> {
    let mut dict = BTreeMap::new();
    while let Some(key_val) = tokens.next() {
        let (key, value) =
            key_val.split_once('=').context("key-value pair delimited by `=` expected")?;
        dict.insert(key, value);
    }
    Ok(dict)
}

#[cfg(test)]
mod tests {
    use super::{ReplayCmd, ReplayMode};

    // These inputs are real mainnet traffic for the given block heights.
    // Each trace contains two chunks in one shard.
    // In combination, they cover an interesting variety of content.
    //  Shard 0: Many receipts, but no function calls.
    //  Shard 1: Empty chunks without txs / receipts.
    //  Shard 2: Some receipts and txs, but no function calls.
    //  Shard 3: Many function calls with lots of storage accesses.
    const INPUT_TRACES: &[&str] = &[
        "75220100-75220101.s0.io_trace",
        "75220100-75220101.s1.io_trace",
        "75220100-75220101.s2.io_trace",
        "75220100-75220101.s3.io_trace",
    ];

    #[test]
    fn test_cache_stats() {
        check_replay_mode(ReplayMode::CacheStats);
    }

    #[test]
    fn test_chunk_cache_stats() {
        check_replay_mode(ReplayMode::ChunkCacheStats);
    }

    #[test]
    fn test_chunk_db_stats() {
        check_replay_mode(ReplayMode::ChunkDbStats);
    }

    #[test]
    fn test_gas_charges() {
        check_replay_mode(ReplayMode::GasCharges);
    }

    #[test]
    fn test_receipt_cache_stats() {
        check_replay_mode(ReplayMode::ReceiptCacheStats);
    }

    #[test]
    fn test_receipt_db_stats() {
        check_replay_mode(ReplayMode::ReceiptDbStats);
    }

    #[track_caller]
    fn check_replay_mode(mode: ReplayMode) {
        for trace_name in INPUT_TRACES {
            let dir = env!("CARGO_MANIFEST_DIR");
            let trace_path = std::path::Path::new(dir).join("res").join(trace_name);
            let cmd = ReplayCmd { trace: trace_path, mode, account: None };
            let mut buffer = Vec::new();
            cmd.run(&mut buffer).unwrap_or_else(|e| {
                panic!("command should not fail for input {trace_name}, failure was {e}")
            });
            let output = String::from_utf8(buffer).unwrap_or_else(|e| {
                panic!("invalid output for input {trace_name}, failure was {e}")
            });
            insta::assert_snapshot!(format!("{mode:?}-{trace_name}"), output);
        }
    }
}
