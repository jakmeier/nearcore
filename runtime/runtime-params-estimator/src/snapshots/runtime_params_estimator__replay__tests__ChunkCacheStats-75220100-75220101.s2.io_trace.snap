---
source: runtime/runtime-params-estimator/src/replay.rs
expression: output
---
apply_transactions shard_id=2 block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
  GET   95 State  
  DB GET               95 requests for a total of    17394 B
  DB SET                0 requests for a total of        0 B
  STORAGE READ          0 requests for a total of        0 B
  STORAGE WRITE         0 requests for a total of        0 B
  TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
  TRIE NODES (host)         162 /  94  (shard-cache/DB)
  SHARD CACHE       63.28% hit rate
  CHUNK CACHE not accessed

apply_transactions shard_id=2 block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
  GET   145 State  
  DB GET              145 requests for a total of    25296 B
  DB SET                0 requests for a total of        0 B
  STORAGE READ          0 requests for a total of        0 B
  STORAGE WRITE         0 requests for a total of        0 B
  TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
  TRIE NODES (host)         307 / 144  (shard-cache/DB)
  SHARD CACHE       68.07% hit rate
  CHUNK CACHE not accessed

top-level:
  GET   4 Block  4 BlockHeader  2 BlockHeight  4 BlockInfo  2 BlockMisc  2 ChunkExtra  2 Chunks  4 EpochInfo  2 IncomingReceipts  2 State  
  DB GET               28 requests for a total of   144614 B
  DB SET                0 requests for a total of        0 B
  STORAGE READ          0 requests for a total of        0 B
  STORAGE WRITE         0 requests for a total of        0 B
  TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
  TRIE NODES (host)           0 /   0  (shard-cache/DB)
  SHARD CACHE not accessed
  CHUNK CACHE not accessed


